### PR TITLE
Fix federated user id resolve logics

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 1]
```diff
@@ -654,35 +656,30 @@ private void storeSessionData(AuthenticationContext context, String sessionConte
...
                     }
+                } catch (UserSessionException e) {
+                    throw new UserSessionException("Error while storing session data for user: "
+                            + user.getLoggableUserId(), e);
+                }
+            } catch (UserIdNotFoundException e) {
+                // If the user id is not available in the user object, or data is not sufficient to resolve it. Hence
+                // the mapping is not stored.
+                if (log.isDebugEnabled()) {
+                    log.debug("A unique user id is not set for the user:" + user.getLoggableUserId()
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:680]:
```suggestion
                    log.debug("A unique user id is not set for the user: " + user.getLoggableUserId()
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 2]
```diff
@@ -214,36 +289,75 @@ public String getAuthenticatedSubjectIdentifier() {
...
 
-    @Override
-    public String getUserId() {
+    public String getUserId() throws UserIdNotFoundException {
 
+        if (this.userId != null) {
+            return userId;
+        }
         // User id can be null sometimes in some flows. Hence trying to resolve it here.
+        userId = resolveUserIdInternal();
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:302]:
```suggestion
        this.userId = resolveUserIdInternal();
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 3]
```diff
@@ -214,36 +289,75 @@ public String getAuthenticatedSubjectIdentifier() {
...
-                        log.debug("User id could not be retrieved for user with username: " + userName + " userstore " +
-                                "domain: " + userStoreDomain + " tenant domain: " + tenantDomain);
-                    }
+                    log.debug("User id could not be resolved for user with username: " + userName + " userstore " +
+                            "domain: " + userStoreDomain + " tenant domain: " + tenantDomain);
                 }
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("User Id retrieved for the user when it is null for user: " + toFullQualifiedUsername() +
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:332]: Fix message

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 4]
```diff
@@ -214,36 +289,75 @@ public String getAuthenticatedSubjectIdentifier() {
...
+
+        this.userId = userId;
+    }
+
+    private String resolveUserIdInternal() {
+
+        String userId = null;
+        if (!isFederatedUser()) {
+            if (log.isDebugEnabled()) {
+                log.debug("User Id retrieved for the user when it is null for user: " + toFullQualifiedUsername() +
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:319]: Fix message

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 5]
```diff
@@ -214,36 +289,75 @@ public String getAuthenticatedSubjectIdentifier() {
...
+                    log.debug("User id could not be resolved for user with username: " + userName + " userstore " +
+                            "domain: " + userStoreDomain + " tenant domain: " + tenantDomain);
                 }
+            }
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("User Id retrieved for the user when it is null for user: " + toFullQualifiedUsername() +
+                        ". Trying to retrieve or generate user id.");
+            }
+            if (federatedIdPName != null && tenantDomain != null && authenticatedSubjectIdentifier != null) {
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:335]: Move the conditions in to `getFederatedUserIdInternal`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 6]
```diff
@@ -152,26 +160,66 @@ public static AuthenticatedUser createLocalAuthenticatedUserFromSubjectIdentifie
...
+                if (userId == null) {
+                    userId = UUID.randomUUID().toString();
+                    UserSessionStore.getInstance()
+                            .storeUserData(userId, this.getAuthenticatedSubjectIdentifier(), tenantId,
+                                    idpId);
+                }
+            } catch (DuplicatedAuthUserException e) {
+
+                String msg = "User authenticated is already persisted. Username: "
+                        + this.getAuthenticatedSubjectIdentifier() + " Tenant Domain:"
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:207]:
```suggestion
                        + this.getAuthenticatedSubjectIdentifier() + " Tenant Domain: "
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 7]
```diff
@@ -152,26 +160,66 @@ public static AuthenticatedUser createLocalAuthenticatedUserFromSubjectIdentifie
...
+                        + this.getTenantDomain() + " IdP: " + this.getFederatedIdPName();
+                log.warn(msg);
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:209]: remove

**Comment by madurangasiriwardena** [comment]
Related to https://github.com/wso2/product-is/issues/8901

**Pull request merged**